testsuite/classlibrary/TestCoreUGens.sc

```diff
@@ -309,124 +318,143 @@ TestCoreUGens : UnitTest {
 				completed = completed + 1;
-				condvar.signalOne;
+				cond.signalOne;
 			});
-			rrand(0.12, 0.35).wait;
+			server.sync;
 		};
 
-		condvar.waitFor(1, { completed == tests.size });
+		testsFinished = cond.waitFor((1.5 * tests.size * testDur).max(minTimeOut), { completed == tests.size });
+		this.assert(testsFinished, "TIMEOUT: ugen_generator_equivalences tests", report: false);
 	}
 
 	test_exact_convergence {
-		var n, v;
-		var condvar = CondVar();
+		var cond = CondVar();
 		var completed = 0;
+		var n, v, tests, testDur, testsFinished;
 
 		// Tests for things that should converge exactly to zero
-		var tests = Dictionary[
+		tests = Dictionary[
+
 			//////////////////////////////////////////
 			// Pan2 amplitude convergence to zero test, unearthed by JH on sc-dev 2009-10-19.
-			"Pan2.ar(ar, , kr) should converge properly to zero when amp set to zero" -> {(Line.ar(1,0,0.2)<=0)*Pan2.ar(BrownNoise.ar, 0, Line.kr(1,0, 0.1)>0).mean},
+			"Pan2.ar(ar, 0, kr) should converge properly to zero when amp set to zero" -> {
+				(Line.ar(1,0,0.2)<=0)*Pan2.ar(BrownNoise.ar, 0, Line.kr(1,0, 0.1)>0).mean
+			},
 
+			// add more tests here
 		];
+
 		server.bootSync;
-		tests.keysValuesDo{|name, func|
-			func.loadToFloatArray(1, server, { |data|
-				this.assertArrayFloatEquals(data, 0, name.quote, within: 0.0, report: true);
+		testDur = defaultNumBlocksToTest * server.options.blockSize / server.sampleRate;
+
+		tests.keysValuesDo{ |text, func|
+			func.loadToFloatArray(testDur, server, { |data|
+				this.assertArrayFloatEquals(data, 0, text.quote, within: 0.0, report: true);
 				completed = completed + 1;
-				condvar.signalOne;
+				cond.signalOne;
 			});
-			rrand(0.05, 0.1).wait;
+			server.sync;
 		};
-
-		condvar.waitFor(1, { completed == tests.size });
+		"timing out after % sec\n".postf(tests.size * testDur);
+		testsFinished = cond.waitFor((1.5 * tests.size * testDur).max(minTimeOut), { completed == tests.size });
+		this.assert(testsFinished, "TIMEOUT: exact_convergence tests", report: false);
 	}
 
+
 	test_muladd {
-		var n, v;
-		var condvar = CondVar();
+		var cond = CondVar();
 		var completed = 0;
+		var n, v, tests, testDur, testsFinished;
 
-		var tests = Dictionary[
-		];
+		tests = Dictionary[];
 		[[\ar,\kr], [2,0,5], [\ar,\kr], [2,0,5], [\ar,\kr], [2,0,5]].allTuples.do{|tup|
-			//tup.postln;
 			tests["%%.madd(%%, %%)".format(*tup)] =
 			"{DC.%(%).madd(DC.%(%), DC.%(%)) - (% * % + %)}".format(*(tup ++ tup[1,3..])).interpret;
 		};
 
 		server.bootSync;
+		testDur = defaultNumBlocksToTest * server.options.blockSize / server.sampleRate;
+
 		tests.keysValuesDo{|name, func|
-			func.loadToFloatArray(0.01, server, { |data|
+			func.loadToFloatArray(testDur, server, { |data|
 				this.assertArrayFloatEquals(data, 0, name.quote, report: true);
 				completed = completed + 1;
-				condvar.signalOne;
+				cond.signalOne;
 			});
+			server.sync;
```

Comment:
@mtmccrea I was just wondering (after #6441) if  you could explain why this `server.sync` is needed? I can't seem to why this passes, but the other approach fails.

Reply:
Ah I think I've figured this out. If you switch to TCP and remove the syncs, you don't need this and performance increases by a factor of 2. I suppose this means the udp messages are getting dropped.

I've also found you get a significant performance increase (circa 5x) in a tight loop by only syncing every 6 iterations (UDP)
```
if (n % 6 == 0) { s.sync };
```

What do you think about switching all these tests over to use TCP and removing the s.sync? It passes on my machine.

---

Sorry for not commenting on this when it was still open!

Reply:
Seems like a great thing to propose in a new PR, so others can weigh in if they know more useful info about UDP vs. TCP behavior. I'm surprised UDP gets dropped on an internal messaging. I'm not sure where the point of failure is vs. TCP... does TCP  just have a larger message queue?

One of the goals of this PR was to model a fairly principled approach to timing tests, which might then find its way into more areas of the test suite. So if we can pin down in some degree of detail why TCP allows higher throughput, we could document it as best practices.

Looks promising!

Reply:
@dyfer linked to @Spacechild1's helpful post [here](https://github.com/supercollider/supercollider/issues/5870#issuecomment-1310965786).

So, if we're running into UDP throughput limits, one solution would be to increase the the socket option's receive buffer size. (not sure if this is compile-time or runtime config... could we just increase this by default in SC??)

In the meantime, `s.sync` seems like the standard option for throttling server-side message backlog.

Reply:
> So, if we're running into UDP throughput limits, one solution would be to increase the the socket option's receive buffer size. 

See https://github.com/supercollider/supercollider/issues/5993 :)

However, I agree with @JordanHendersonMusic that if you want to have reliable communication, TCP is the way to go!

>  So if we can pin down in some degree of detail why TCP allows higher throughput, we could document it as best practices.

With UDP you need to artificially throttle the traffice with `s.sync` or `wait` to prevent socket receive buffer overruns. With TCP this is not necessary, which leads to the curious situation that in some situations you get higher throughput, even though TCP is the slower protocol.